File: perl5/db_maintainer_28dc4d.pl
#!/usr/bin/perl
# db_maintainer.py — portiert nach perl5
# Quelle: python, OpenClaw@gateway2:skills/db-maintainer/scripts/db_maintainer.py
# Erzeugt: 2026-08-19 durch ABSTRACTIONS_MANAGER.py

use strict;
use warnings;
use utf8;
use JSON;
use File::Path qw(make_path);
use File::Copy;
use File::Find;
use Digest::MD5 qw(md5_hex);
use POSIX qw(strftime);

=head1 Database Maintainer Sub-Agent

Automated database maintenance with 30min checks, hourly backups (3 days retention),
band tree command execution for important/openclaw-tree.txt

=cut

my $WORKSPACE = "/home/openclaw/.openclaw/workspace";
my $DB_DIR = "$WORKSPACE/db";
my $BACKUP_DIR = "$DB_DIR/backups";
my $LOG_DIR = "$WORKSPACE/logs/db-maintainer";
my $IMPORTANT_DIR = "$WORKSPACE/important";

# Create directories
make_path($BACKUP_DIR, {verbose => 0}) or die "Couldn't create $BACKUP_DIR: $!";
make_path($LOG_DIR, {verbose => 0}) or die "Couldn't create $LOG_DIR: $!";

package Logger;

sub new {
    my $class = shift;
    my $self = {};
    bless $self, $class;
    return $self;
}

sub log {
    my ($self, $level, $message) = @_;
    my $timestamp = strftime('%Y-%m-%d %H:%M:%S', localtime);
    my $line = "[$timestamp] [$level] $message\n";
    print $line;
    my $today = strftime('%Y-%m-%d', localtime);
    my $log_file = "$LOG_DIR/$today.log";
    open(my $fh, '>>', $log_file) or die "Could not open file '$log_file': $!";
    print $fh $line;
    close $fh;
}

sub info { $_[0]->log('INFO', $_[1]); }
sub warn { $_[0]->log('WARN', $_[1]); }
sub error { $_[0]->log('ERROR', $_[1]); }

package DatabaseMaintainer;

sub new {
    my $class = shift;
    my $self = {
        logger => Logger->new(),
        state_file => "$DB_DIR/maintainer_state.json",
        retention_days => 3
    };
    bless $self, $class;
    return $self;
}

sub load_state {
    my $self = shift;
    if (-f $self->{state_file}) {
        open(my $fh, '<', $self->{state_file}) or die "Could not open file '$self->{state_file}': $!";
        my $json_text = do { local $/; <$fh> };
        close $fh;
        return decode_json($json_text);
    }
    return {
        last_check => undef,
        last_backup => undef,
        last_tree_update => undef,
        file_hashes => {}
    };
}

sub save_state {
    my ($self, $state) = @_;
    open(my $fh, '>', $self->{state_file}) or die "Could not open file '$self->{state_file}': $!";
    print $fh encode_json($state);
    close $fh;
}

sub get_file_hash {
    my ($self, $filepath) = @_;
    if (-f $filepath) {
        open(my $fh, '<', $filepath) or return undef;
        binmode($fh);
        my $data = do { local $/; <$fh> };
        close $fh;
        return md5_hex($data);
    }
    return undef;
}

sub run_tree_command {
    my $self = shift;
    eval {
        local $SIG{ALRM} = sub { die "timeout\n" };
        alarm(60);
        my @cmd = ('tree', '-a', '-L', '6', $WORKSPACE);
        open(my $pipe, '-|', @cmd) or die "Cannot run tree command: $!";
        my $output = do { local $/; <$pipe> };
        close $pipe;
        alarm(0);
        $self->{logger}->info("tree -a -L 6 erfolgreich ausgeführt");
        return $output;
    };
    if ($@) {
        if ($@ eq "timeout\n") {
            $self->{logger}->error("tree command timed out");
        } else {
            $self->{logger}->error("tree command fehlgeschlagen: $@");
        }
        return undef;
    }
}

sub update_tree_file {
    my ($self, $tree_output) = @_;
    return 0 unless defined $tree_output;
    
    my $tree_file = "$IMPORTANT_DIR/openclaw-tree.txt";
    my $header = "# OpenClaw Workspace Tree\n";
    $header .= "# Generiert: " . strftime('%Y-%m-%dT%H:%M:%S', localtime) . "\n";
    $header .= "# Befehl: tree -a -L 6 $WORKSPACE\n";
    $header .= "# Diese Datei wird automatisch von db-maintainer aktualisiert\n\n";
    
    eval {
        open(my $fh, '>', $tree_file) or die "Cannot write to $tree_file: $!";
        print $fh $header;
        print $fh $tree_output;
        close $fh;
        $self->{logger}->info("openclaw-tree.txt aktualisiert: $tree_file");
        return 1;
    };
    if ($@) {
        $self->{logger}->error("Fehler beim Schreiben von openclaw-tree.txt: $@");
        return 0;
    }
}

sub scan_documentations {
    my $self = shift;
    my @docs;
    my @patterns = ("$WORKSPACE/*.md", "$WORKSPACE/**/*.md");
    
    find(sub {
        return unless /\.md$/ && -f $_ && !-l $_;
        my $full_path = $File::Find::name;
        return if $full_path =~ m{/db/backups/} || $full_path =~ m{/node_modules/};
        my $relative_path = $full_path;
        $relative_path =~ s{^\Q$WORKSPACE\E/?}{};
        push @docs, {
            path => $relative_path,
            hash => $self->get_file_hash($full_path),
            mtime => (stat($_))[9]
        };
    }, $WORKSPACE);
    
    return \@docs;
}

sub check_for_changes {
    my $self = shift;
    my $state = $self->load_state();
    my $current_docs = $self->scan_documentations();
    
    my @changes;
    my %current_hashes;
    
    for my $doc (@$current_docs) {
        my $path = $doc->{path};
        $current_hashes{$path} = $doc->{hash};
        
        if (!exists $state->{file_hashes}{$path}) {
            push @changes, "NEW: $path";
        } elsif ($state->{file_hashes}{$path} ne $doc->{hash}) {
            push @changes, "CHANGED: $path";
        }
    }
    
    # Check for deleted files
    for my $old_path (keys %{$state->{file_hashes}}) {
        if (!exists $current_hashes{$old_path}) {
            push @changes, "DELETED: $old_path";
        }
    }
    
    return (\@changes, \%current_hashes);
}

sub update_databases {
    my $self = shift;
    eval {
        local $SIG{ALRM} = sub { die "timeout\n" };
        alarm(60);
        system("python3", "$WORKSPACE/scripts/update_docs_db.py");
        alarm(0);
        if ($? == 0) {
            $self->{logger}->info("docs.db aktualisiert");
            return 1;
        } else {
            $self->{logger}->error("DB-Update fehlgeschlagen: $?");
            return 0;
        }
    };
    if ($@) {
        $self->{logger}->error("DB-Update Exception: $@");
        return 0;
    }
}

sub update_tree_db_v2 {
    my $self = shift;
    eval {
        local $SIG{ALRM} = sub { die "timeout\n" };
        alarm(120);
        system("python3", "$WORKSPACE/scripts/tree_indexer_v2.py");
        alarm(0);
        if ($? == 0) {
            $self->{logger}->info("tree.db v2 aktualisiert");
            return 1;
        } else {
            $self->{logger}->error("Tree-DB v2 fehlgeschlagen: $?");
            return 0;
        }
    };
    if ($@) {
        $self->{logger}->error("Tree-DB v2 Exception: $@");
        return 0;
    }
}

sub create_backup {
    my $self = shift;
    my $timestamp = strftime('%Y-%m-%d_%H-%M', localtime);
    
    for my $db_name ('docs.db', 'tree.db') {
        my $source = "$DB_DIR/$db_name";
        if (-f $source) {
            my $backup_name = "${timestamp}_${db_name}.bak";
            my $backup_path = "$BACKUP_DIR/$backup_name";
            copy($source, $backup_path) or die "Copy failed: $!";
            $self->{logger}->info("Backup erstellt: $backup_name");
        }
    }
    
    return $timestamp;
}

sub cleanup_old_backups {
    my $self = shift;
    my $cutoff = time() - ($self->{retention_days} * 24 * 60 * 60);
    my $deleted = 0;
    
    for my $db_name ('docs.db', 'tree.db') {
        opendir(my $dh, $BACKUP_DIR) or die "Cannot opendir $BACKUP_DIR: $!";
        my @backups = grep { /\.bak$/ && /_$db_name\.bak$/ } readdir($dh);
        closedir $dh;
        
        for my $backup (@backups) {
            # Extract date from filename (Format: YYYY-MM-DD_HH-MM)
            if ($backup =~ /^(\d{4}-\d{2}-\d{2})_(\d{2}-\d{2})/) {
                my $date_part = $1;
                my $time_part = $2;
                my $backup_time_str = "${date_part}_${time_part}";
                my $backup_timestamp = _parse_datetime($backup_time_str);
                
                if ($backup_timestamp && $backup_timestamp < $cutoff) {
                    unlink("$BACKUP_DIR/$backup") or warn "Could not delete $backup: $!";
                    $deleted++;
                    $self->{logger}->info("Altes Backup gelöscht: $backup");
                }
            } else {
                $self->{logger}->warn("Konnte Backup-Datum nicht parsen: $backup");
            }
        }
    }
    
    if ($deleted == 0) {
        $self->{logger}->info("Keine alten Backups zum Löschen");
    } else {
        $self->{logger}->info("$deleted alte Backups gelöscht (< 3 Tage)");
    }
}

sub _parse_datetime {
    my $datetime_str = shift;
    if ($datetime_str =~ /^(\d{4})-(\d{2})-(\d{2})_(\d{2})-(\d{2})$/) {
        my ($year, $month, $day, $hour, $minute) = ($1, $2, $3, $4, $5);
        use Time::Local;
        return timelocal(0, $minute, $hour, $day, $month-1, $year);
    }
    return undef;
}

sub run_cycle {
    my $self = shift;
    $self->{logger}->info("=" x 60);
    $self->{logger}->info("DB MAINTAINER CYCLE START");
    $self->{logger}->info("=" x 60);
    
    my $state = $self->load_state();
    
    # 1. Run tree command and write to openclaw-tree.txt
    $self->{logger}->info("Führe tree -a -L 8 aus...");
    my $tree_output = $self->run_tree_command();
    if ($tree_output) {
        $self->update_tree_file($tree_output);
        $state->{last_tree_update} = strftime('%Y-%m-%dT%H:%M:%S', localtime);
    }
    
    # 2. Update tree.db (internal v2)
    $self->{logger}->info("Aktualisiere tree.db v2...");
    $self->update_tree_db_v2();
    
    # 3. Check for changes
    $self->{logger}->info("Prüfe auf Dokumentations-Änderungen...");
    my ($changes, $current_hashes) = $self->check_for_changes();
    
    if (@$changes) {
        $self->{logger}->info(scalar(@$changes) . " Änderungen gefunden:");
        my $count = 0;
        for my $change (@$changes) {
            last if $count++ >= 10;
            $self->{logger}->info("  - $change");
        }
        if (@$changes > 10) {
            $self->{logger}->info("  ... und " . (scalar(@$changes) - 10) . " weitere");
        }
        
        # 4. Update docs.db
        $self->{logger}->info("Aktualisiere docs.db...");
        if ($self->update_databases()) {
            $state->{last_check} = strftime('%Y-%m-%dT%H:%M:%S', localtime);
            $state->{file_hashes} = $current_hashes;
        }
    } else {
        $self->{logger}->info("Keine Dokumentations-Änderungen gefunden");
    }
    
    # 5. Check if backup is due (hourly)
    my $do_backup = 1;
    if ($state->{last_backup}) {
        my $last_backup_time = _parse_datetime($state->{last_backup});
        if ($last_backup_time) {
            my $now = time();
            $do_backup = ($now - $last_backup_time) >= 3600; # 1 hour in seconds
        }
    }
    
    if ($do_backup) {
        $self->{logger}->info("Erstelle stündliches Backup...");
        my $timestamp = $self->create_backup();
        $state->{last_backup} = strftime('%Y-%m-%dT%H:%M:%S', localtime);
        
        # 6. Clean up old backups (3 day retention)
        $self->{logger}->info("Räume alte Backups auf (3 Tage Retention)...");
        $self->cleanup_old_backups();
    } else {
        $self->{logger}->info("Backup nicht nötig (letztes < 1h)");
    }
    
    $self->save_state($state);
    
    $self->{logger}->info("=" x 60);
    $self->{logger}->info("DB MAINTAINER CYCLE END");
    $self->{logger}->info("=" x 60);
}

sub main {
    my $maintainer = DatabaseMaintainer->new();
    
    eval {
        $maintainer->run_cycle();
    };
    if ($@) {
        $maintainer->{logger}->error("CRITICAL ERROR: $@");
        exit 1;
    }
}

main() unless caller;

File: javascript/db_maintainer_28dc4d.js
#!/usr/bin/env node
// db_maintainer.py — portiert nach javascript
// Quelle: python, OpenClaw@gateway2:skills/db-maintainer/scripts/db_maintainer.py
// Erzeugt: 2026-08-19 durch ABSTRACTIONS_MANAGER.py

/**
 * Database Maintainer Sub-Agent
 * Automated database maintenance with 30min checks, hourly backups (3 days retention),
 * band tree command execution for important/openclaw-tree.txt
 */

const fs = require('fs');
const path = require('path');
const crypto = require('crypto');
const { spawnSync } = require('child_process');

const WORKSPACE = '/home/openclaw/.openclaw/workspace';
const DB_DIR = path.join(WORKSPACE, 'db');
const BACKUP_DIR = path.join(DB_DIR, 'backups');
const LOG_DIR = path.join(WORKSPACE, 'logs', 'db-maintainer');
const IMPORTANT_DIR = path.join(WORKSPACE, 'important');

// Verzeichnisse erstellen
[BACKUP_DIR, LOG_DIR].forEach(dir => {
    if (!fs.existsSync(dir)) {
        fs.mkdirSync(dir, { recursive: true });
    }
});

class Logger {
    /** Einfacher Logger mit Datei-Ausgabe */
    
    constructor() {
        const today = new Date().toISOString().split('T')[0];
        this.logFile = path.join(LOG_DIR, `${today}.log`);
    }
    
    log(level, message) {
        const timestamp = new Date().toISOString().replace('T', ' ').substring(0, 19);
        const line = `[${timestamp}] [${level}] ${message}`;
        console.log(line);
        fs.appendFileSync(this.logFile, line + '\n');
    }
    
    info(msg) { this.log('INFO', msg); }
    warn(msg) { this.log('WARN', msg); }
    error(msg) { this.log('ERROR', msg); }
}

class DatabaseMaintainer {
    constructor() {
        this.logger = new Logger();
        this.stateFile = path.join(DB_DIR, 'maintainer_state.json');
        this.retentionDays = 3; // 3 Tage Backup-Aufbewahrung
    }
    
    loadState() {
        /** Lädt letzten Check-Zustand */
        if (fs.existsSync(this.stateFile)) {
            const data = fs.readFileSync(this.stateFile, 'utf8');
            return JSON.parse(data);
        }
        return { last_check: null, last_backup: null, last_tree_update: null, file_hashes: {} };
    }
    
    saveState(state) {
        /** Speichert aktuellen Zustand */
        fs.writeFileSync(this.stateFile, JSON.stringify(state, null, 2));
    }
    
    getFileHash(filepath) {
        /** Berechnet MD5-Hash einer Datei */
        try {
            const data = fs.readFileSync(filepath);
            return crypto.createHash('md5').update(data).digest('hex');
        } catch {
            return null;
        }
    }
    
    runTreeCommand() {
        /** Führt tree -a -L 6 auf workspace aus und gibt Ergebnis zurück */
        try {
            const result = spawnSync('tree', ['-a', '-L', '6', WORKSPACE], {
                encoding: 'utf8',
                timeout: 60000
            });
            
            if (result.status === 0) {
                this.logger.info("tree -a -L 6 erfolgreich ausgeführt");
                return result.stdout;
            } else {
                this.logger.error(`tree command fehlgeschlagen: ${result.stderr}`);
                return null;
            }
        } catch (e) {
            this.logger.error(`tree command Exception: ${e.message}`);
            return null;
        }
    }
    
    updateTreeFile(treeOutput) {
        /** Schreibt tree-output in important/openclaw-tree.txt */
        if (!treeOutput) {
            return false;
        }
        
        const treeFile = path.join(IMPORTANT_DIR, 'openclaw-tree.txt');
        
        // Header mit Timestamp
        const header = `# OpenClaw Workspace Tree
# Generiert: ${new Date().toISOString()}
# Befehl: tree -a -L 6 ${WORKSPACE}
# Diese Datei wird automatisch von db-maintainer aktualisiert

`;
        
        try {
            fs.writeFileSync(treeFile, header + treeOutput);
            this.logger.info(`openclaw-tree.txt aktualisiert: ${treeFile}`);
            return true;
        } catch (e) {
            this.logger.error(`Fehler beim Schreiben von openclaw-tree.txt: ${e.message}`);
            return false;
        }
    }
    
    scanDocumentations() {
        /** Scannt alle .md Dateien auf Änderungen */
        const docs = [];
        const walkDir = (dir) => {
            const entries = fs.readdirSync(dir, { withFileTypes: true });
            for (const entry of entries) {
                const fullPath = path.join(dir, entry.name);
                const relativePath = path.relative(WORKSPACE, fullPath);
                
                // Skip backup and node_modules directories
                if (relativePath.includes('db/backups') || relativePath.includes('node_modules')) {
                    continue;
                }
                
                if (entry.isFile() && entry.name.endsWith('.md')) {
                    const stat = fs.statSync(fullPath);
                    docs.push({
                        path: relativePath,
                        hash: this.getFileHash(fullPath),
                        mtime: stat.mtimeMs
                    });
                } else if (entry.isDirectory()) {
                    walkDir(fullPath);
                }
            }
        };
        
        walkDir(WORKSPACE);
        return docs;
    }
    
    checkForChanges() {
        /** Prüft auf Änderungen seit letztem Lauf */
        const state = this.loadState();
        const currentDocs = this.scanDocumentations();
        
        const changes = [];
        const currentHashes = {};
        
        for (const doc of currentDocs) {
            const docPath = doc.path;
            currentHashes[docPath] = doc.hash;
            
            if (!(docPath in state.file_hashes)) {
                changes.push(`NEW: ${docPath}`);
            } else if (state.file_hashes[docPath] !== doc.hash) {
                changes.push(`CHANGED: ${docPath}`);
            }
        }
        
        // Prüfe auf gelöschte Dateien
        for (const oldPath in state.file_hashes) {
            if (!(oldPath in currentHashes)) {
                changes.push(`DELETED: ${oldPath}`);
            }
        }
        
        return changes, currentHashes;
    }
    
    updateDatabases() {
        /** Führt DB-Update-Scripts aus */
        try {
            const scriptPath = path.join(WORKSPACE, 'scripts', 'update_docs_db.py');
            const result = spawnSync('python3', [scriptPath], {
                encoding: 'utf8',
                timeout: 60000
            });
            
            if (result.status === 0) {
                this.logger.info("docs.db aktualisiert");
                return true;
            } else {
                this.logger.error(`DB-Update fehlgeschlagen: ${result.stderr}`);
                return false;
            }
        } catch (e) {
            this.logger.error(`DB-Update Exception: ${e.message}`);
            return false;
        }
    }
    
    updateTreeDbV2() {
        /** Führt tree_indexer_v2.py aus */
        try {
            const scriptPath = path.join(WORKSPACE, 'scripts', 'tree_indexer_v2.py');
            const result = spawnSync('python3', [scriptPath], {
                encoding: 'utf8',
                timeout: 120000
            });
            
            if (result.status === 0) {
                this.logger.info("tree.db v2 aktualisiert");
                return true;
            } else {
                this.logger.error(`Tree-DB v2 fehlgeschlagen: ${result.stderr}`);
                return false;
            }
        } catch (e) {
            this.logger.error(`Tree-DB v2 Exception: ${e.message}`);
            return false;
        }
    }
    
    createBackup() {
        /** Erstellt Backup beider Datenbanken */
        const timestamp = new Date().toISOString().replace(/[:T]/g, '-').substring(0, 16);
        
        ['docs.db', 'tree.db'].forEach(dbName => {
            const source = path.join(DB_DIR, dbName);
            if (fs.existsSync(source)) {
                const backupName = `${timestamp}_${dbName}.bak`;
                const backupPath = path.join(BACKUP_DIR, backupName);
                fs.copyFileSync(source, backupPath);
                this.logger.info(`Backup erstellt: ${backupName}`);
            }
        });
        
        return timestamp;
    }
    
    cleanupOldBackups() {
        /** Löscht Backups älter als 3 Tage */
        const cutoff = new Date(Date.now() - this.retentionDays * 24 * 60 * 60 * 1000);
        let deleted = 0;
        
        ['docs.db', 'tree.db'].forEach(dbName => {
            const backups = fs.readdirSync(BACKUP_DIR)
                .filter(file => file.endsWith(`_${dbName}.bak`))
                .map(file => path.join(BACKUP_DIR, file));
            
            backups.forEach(backup => {
                try {
                    const filename = path.basename(backup);
                    const dateStr = filename.split('_')[0];
                    const timeStr = filename.split('_')[1];
                    const backupTime = new Date(`${dateStr}T${timeStr.replace('-', ':')}`);
                    
                    if (backupTime < cutoff) {
                        fs.unlinkSync(backup);
                        deleted++;
                        this.logger.info(`Altes Backup gelöscht: ${filename}`);
                    }
                } catch {
                    this.logger.warn(`Konnte Backup-Datum nicht parsen: ${path.basename(backup)}`);
                }
            });
        });
        
        if (deleted === 0) {
            this.logger.info("Keine alten Backups zum Löschen");
        } else {
            this.logger.info(`${deleted} alte Backups gelöscht (< 3 Tage)`);
        }
    }
    
    runCycle() {
        /** Ein kompletter Wartungszyklus */
        this.logger.info("=".repeat(60));
        this.logger.info("DB MAINTAINER CYCLE START");
        this.logger.info("=".repeat(60));
        
        const state = this.loadState();
        
        // 1. Tree-Befehl ausführen und in openclaw-tree.txt schreiben
        this.logger.info("Führe tree -a -L 8 aus...");
        const treeOutput = this.runTreeCommand();
        if (treeOutput) {
            this.updateTreeFile(treeOutput);
            state.last_tree_update = new Date().toISOString();
        }
        
        // 2. tree.db aktualisieren (intern v2)
        this.logger.info("Aktualisiere tree.db v2...");
        this.updateTreeDbV2();
        
        // 3. Änderungen prüfen
        this.logger.info("Prüfe auf Dokumentations-Änderungen...");
        const [changes, currentHashes] = this.checkForChanges();
        
        if (changes.length > 0) {
            this.logger.info(`${changes.length} Änderungen gefunden:`);
            changes.slice(0, 10).forEach(change => {
                this.logger.info(`  - ${change}`);
            });
            if (changes.length > 10) {
                this.logger.info(`  ... und ${changes.length - 10} weitere`);
            }
            
            // 4. docs.db aktualisieren
            this.logger.info("Aktualisiere docs.db...");
            if (this.updateDatabases()) {
                state.last_check = new Date().toISOString();
                state.file_hashes = currentHashes;
            }
        } else {
            this.logger.info("Keine Dokumentations-Änderungen gefunden");
        }
        
        // 5. Prüfe ob Backup fällig (stündlich)
        const lastBackup = state.last_backup;
        let doBackup = false;
        
        if (lastBackup) {
            const lastBackupTime = new Date(lastBackup);
            doBackup = (Date.now() - lastBackupTime.getTime()) >= 60 * 60 * 1000; // 1 hour
        } else {
            doBackup = true;
        }
        
        if (doBackup) {
            this.logger.info("Erstelle stündliches Backup...");
            const timestamp = this.createBackup();
            state.last_backup = new Date().toISOString();
            
            // 6. Alte Backups aufräumen (3 Tage Retention)
            this.logger.info("Räume alte Backups auf (3 Tage Retention)...");
            this.cleanupOldBackups();
        } else {
            this.logger.info("Backup nicht nötig (letztes < 1h)");
        }
        
        this.saveState(state);
        
        this.logger.info("=".repeat(60));
        this.logger.info("DB MAINTAINER CYCLE END");
        this.logger.info("=".repeat(60));
    }
}

function main() {
    /** Hauptfunktion */
    const maintainer = new DatabaseMaintainer();
    
    try {
        maintainer.runCycle();
    } catch (e) {
        maintainer.logger.error(`CRITICAL ERROR: ${e.message}`);
        process.exit(1);
    }
}

main();
